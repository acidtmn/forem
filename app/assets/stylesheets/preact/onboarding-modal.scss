@import 'variables';
@import 'mixins';

#onboarding-container {
  min-height: calc(100vh - 350px);
}

.pages-onboarding {
  min-height: initial;
}

// Modal Box Card
.onboarding-main {
  @include themeable(background-color, theme-container-background, $white);
  @include themeable(border-color, theme-container-border, black);
  align-items: center;
  border-radius: 3px;
  border: 2px solid black;
  box-shadow: 4px 4px black;
  display: flex;
  flex-direction: column;
  font-size: 16px;
  justify-content: center;
  min-height: 460px;
  width: 740px;
  z-index: 1;
}

// Modal Inner Content Spacing
.onboarding-content {
  box-sizing: border-box;
  flex: 1 0 auto;
  padding: 32px 48px 48px;
  width: 100%;

  @media screen and (min-width: 800px) {
  }
}

// Modal Header
.onboarding-content-header {
  margin-bottom: 24px;

  .title,
  .subtitle {
    margin: 0;
  }

  .title {
    font-weight: 900;
    text-align: left;
    margin-bottom: 4px;
  }

  .subtitle {
    color: #363d44;
    font-size: 20px;
    font-weight: 500;
    letter-spacing: normal;
    line-height: 150%;
  }
}

// Modal Navigation
.onboarding-navigation {
  align-self: flex-end;
  flex-shrink: 0;
  width: 100%;
  border-bottom: 2px solid black;

  .navigation-content {
    display: flex;
    justify-content: space-between;
    padding: 16px;
  }
}

// Modal Selection Status
.onboarding-selection-status {
  align-self: flex-end;
  flex-shrink: 0;
  width: 100%;
  border-top: 2px solid black;

  .selection-status-content {
    display: block;
    padding: 16px;
    text-align: center;
  }
}

// Button Styles
.onboarding-body button {
  -webkit-appearance: none;
  background-color: #3b49df;
  border-radius: 3px;
  border: none;
  box-sizing: border-box;
  color: white;
  font-size: 16px;
  font-weight: 500;
  line-height: 150%;
  padding: 0.5em 1.5em;
  text-align: center;
}

// Tag Styles
.tag-container {
  box-sizing: border-box;
  column-gap: 1em;
  display: grid;
  grid-gap: 1em;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: auto;
  width: 100%;

  .tag-item {
    background-color: #eef0f1; // smoke-10
    border: 2px solid #eef0f1;
    border-radius: 3px;
    display: grid;
    row-gap: 1em;
    height: 120px;
    padding: 16px;

    &:hover {
      border-color: black;
      background: white;
      cursor: pointer;

      .tag-item-selector {
        display: block;
      }
    }

    .tag-topic-name {
      font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier,
        monospace;
      font-size: 16px;
      font-weight: 400;
      margin: 0;
    }

    .tag-item-selector {
      display: none;
      padding: 0;
    }
  }
}

.onboarding-tags {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 24px;
  width: 100%;

  &__button {
    font-size: 16px;
    font-weight: 500;
    line-height: 24px;
    padding: 0.5em;
    border: 0;
    margin: 0;
    display: block;
    text-align: center;
    position: absolute;
    left: 16px;
    right: 16px;
    width: calc(100% - 32px);
    bottom: 16px;
    border-radius: 3px;
    transition: all ease-in-out 0.1s;
    transform: translateY(60px);
    overflow: hidden;
    $button: &;

    &-default {
      display: inline-block;
    }

    &-alt {
      display: none;
    }
  }

  &__item {
    border-radius: 3px;
    position: relative;
    overflow: hidden;
    height: 150px;
    $item: &;

    &__inner {
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      z-index: 2;
      background: #eef0f1;
      transition: all ease-in-out 0.1s;
      padding: 16px;
    }

    &:hover {
      #{$item}__inner {
        background: transparent;
      }

      .onboarding-tags__button {
        transform: translateY(0);
        &:hover {
          transform: translateY(-2px);
        }
      }
    }

    &--selected {
      #{$item}__inner {
        background: transparent;
      }

      .onboarding-tags__button {
        transform: translateY(0);
        $button: &;

        &:hover {
          .onboarding-tags__button-default {
            display: none;
          }

          .onboarding-tags__button-alt {
            display: inline-block;
          }
        }
      }
    }
  }
}

// Onboarding Intro

.onboarding-main.introduction {
  width: 578px;

  .onboarding-content {
    background: #08090a;
    color: white;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 48px;
  }

  figure {
    margin: 0 0 8px 0;

    .sticker-logo {
      height: 48px;
      width: 48px;
      transform: rotate(-30deg);
      padding: 10px;
      vertical-align: middle;
    }
  }

  .introduction-title,
  .introduction-subtitle {
    margin: 0;
  }

  .introduction-title {
    font-size: 48px;
    font-weight: 900;
    line-height: 125%;
  }

  .introduction-subtitle {
    font-size: 24px;
    font-weight: 400;
    line-height: 150%;
  }

  .navigation-content {
    padding: 32px;
    justify-content: center;
  }
}

// Next steps

// Background
.onboarding-body {
  background-image: url('/assets/onboarding-background-white.png');
  position: fixed;
  margin-top: 50px;
  bottom: 0;
  height: 100vh;
  width: 100%;
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
  line-height: 200%;
  z-index: 11;
  margin: 0;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  /* Center the image */
  background-position: center;
  /* Do not repeat the image */
  background-repeat: no-repeat;

  &:before {
    display: block;
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: rgba(0, 0, 0, 0.65);
    height: 100%;
    width: 100%;
  }

  a {
    @include themeable(color, theme-anchor-color, #2e3b9e);
  }

  .background-image {
    z-index: -1;
    position: absolute;
    top: 60px;
    left: 0px;
  }

  .badge-image {
    position: absolute;
  }

  .onboarding-what-next {
    display: flex;
    flex-direction: column;
    height: 250px;
    overflow: auto;

    @media screen and (min-width: 600px) {
      flex-direction: row;
    }

    a {
      background: #d9e7ff;
      padding: 15px 4px 0px 12px;
      height: 140px;
      margin: 4px auto;
      border-radius: 3px;
      width: 350px;
      display: block;
      line-height: 1.2em;
      font-weight: bold;
      position: relative;
      text-align: center;

      @media screen and (min-width: 600px) {
        height: 220px;
        margin: 4px 4px;
        text-align: left;
        padding: 20px 4px 0px 12px;
      }

      .whatnext-emoji {
        display: none;
        text-align: center;
        position: absolute;
        bottom: 22px;
        left: 0;
        right: 0;
        font-size: 50px;

        @media screen and (min-width: 600px) {
          display: block;
        }
      }
    }
  }

  .onboarding-previous-location {
    text-align: center;
    margin-top: 10px;
    font-weight: bold;
    display: none;

    @media screen and (min-width: 600px) {
      display: block;
    }

    code {
      font-size: 0.6em;
      margin-top: -10px;
      display: block;
      font-weight: 400;
    }
  }

  @media screen and (max-width: 800px) {
    .onboarding-main {
      width: 100%;
      height: 100%;
      background-image: url('/assets/mobile-onboarding-background.png');
      background-size: cover;
      /* Center the image */
      background-position: center;
      /* Do not repeat the image */
      background-repeat: no-repeat;
      border: none;
      color: $black;
    }

    .onboarding-content {
      flex: 0 0 auto;

      h1,
      h2 {
        margin-top: initial;
      }
    }

    span {
      display: block;
    }
  }

  @media screen and (max-width: 650px) {
    .onboarding-main {
      font-size: 18px;
    }

    button {
      width: 100px;
      font-size: 20px;
    }
  }

  @media screen and (max-width: 450px) {
    .onboarding-main {
      font-size: 16px;
    }

    button {
      width: 100px;
      font-size: 18px;
    }
  }

  @media screen and (max-width: 330px) {
    button {
      width: 80px;
      font-size: 14px;
    }

    .sticker-logo {
      height: 40px;
      width: 40px;
    }

    h1 {
      font-size: 20px;
    }
  }
}

// Follow user
.onboarding-body {
  .content-row {
    align-items: center;
    color: #202428; // smoke-90
    display: flex;
    padding: 8px 16px;
    text-align: left;
    width: 100%;

    &:hover {
      background-color: #eef0f1 !important; //smoke-10
    }
  }

  .user-avatar-container {
    margin: 0 16px 0 0;
  }

  .user-avatar {
    border-radius: 100%;
    height: 48px;
    width: 48px;
    object-fit: cover;
    border: 1px solid #eef0f1; //smoke-10
  }

  .user-name,
  .user-summary {
    margin: 0;
  }

  .user-following-status {
    margin-left: 32px;
  }
}
