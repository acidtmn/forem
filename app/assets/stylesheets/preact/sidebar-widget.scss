@import 'variables';

.sidebar-additional {
  .widget {
    .widget-suggested-follows-container {
      h4 {
        margin: 0;
      }
      .widget-body {
        .widget-list-item__suggestions {
          display: flex;
          align-items: center;

          .widget-list-item__avatar {
            width: 48px;
            height: 48px;

            img {
              width: 100%;
              height: 100%;
              border-radius: 100%;
            }
          }

          .widget-list-item__content {
            flex: 1 auto;
            padding-left: 8px;
            h5 {
              all: unset;
              display: block;
              margin-bottom: 4px;
            }
          }

          .widget-list-item__follow-button {
            border: 2px solid transparent;
            border-radius: 3px;
            padding: 2px 8px;
            text-align: center;
            background-color: $green;
            color: $black;
            font-family: $helvetica-condensed;
            font-weight: 900;
            font-size: 0.8em;
          }
        }
      }
    }
  }
}
