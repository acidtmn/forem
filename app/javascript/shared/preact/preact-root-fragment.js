/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  replaceNode = [].concat(replaceNode);
  var s = replaceNode[replaceNode.length-1].nextSibling;
  function insert(c, r) { parent.insertBefore(c, r || s); }
  return parent.__k = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: insert,
    appendChild: insert,
    removeChild: function(c) { parent.removeChild(c); }
  };
}