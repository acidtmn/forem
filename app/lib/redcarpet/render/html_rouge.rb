require "rouge/plugins/redcarpet"

module Redcarpet
  module Render
    class HTMLRouge < HTML
      include Rouge::Plugins::Redcarpet

      # overrided method to add line number support
      def block_code(code, language, opts = {})
        lexer =
          begin
            Rouge::Lexer.find_fancy(language, code)
          rescue Rouge::Guesser::Ambiguous => e
            e.alternatives.first
          end
        lexer ||= Rouge::Lexers::PlainText

        # XXX HACK: Redcarpet strips hard tabs out of code blocks,
        # so we assume you're not using leading spaces that aren't tabs,
        # and just replace them here.
        if lexer.tag == "make"
          code.gsub! %r{^    }, "\t"
        end

        # append default css_class to block_code
        unless opts.key? "css_class"
          opts["css_class"] = "highlight #{lexer.tag}"
        end

        formatter = Rouge::Formatters::HTMLLegacy.new(opts)
        formatter.format(lexer.lex(code))
      end

      def link(link, _title, content)
        # Probably not the best fix but it does it's job of preventing
        # a nested links.
        return nil if /<a\s.+\/a>/.match?(content)

        link_attributes = ""
        @options[:link_attributes]&.each do |attribute, value|
          link_attributes += %( #{attribute}="#{value}")
        end
        %(<a href="#{link}"#{link_attributes}>#{content}</a>)
      end

      def header(title, header_number)
        anchor_link = slugify(title)
        <<~HEREDOC
          <h#{header_number}>
            <a name="#{anchor_link}" href="##{anchor_link}" class="anchor">
            </a>
            #{title}
          </h#{header_number}>
        HEREDOC
      end

      private

      def slugify(string)
        stripped_string = ActionView::Base.full_sanitizer.sanitize string
        stripped_string.downcase.gsub(EmojiRegex::Regex, "").strip.gsub(/[[:punct:]]/u, "").gsub(/\s+/, "-")
      end
    end
  end
end
