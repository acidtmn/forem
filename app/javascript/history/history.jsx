import { h, Component } from 'preact';
import { PropTypes } from 'preact-compat';
import debounce from 'lodash.debounce';
import setupAlgoliaIndex from '../src/utils/algolia';

import { ItemListLoadMoreButton } from '../src/components/ItemList/ItemListLoadMoreButton';
import { ItemListTags } from '../src/components/ItemList/ItemListTags';

export class History extends Component {
  constructor(props) {
    super(props);

    this.handleTyping = debounce(this.handleTyping.bind(this), 300, {
      leading: true,
    });

    const { availableTags } = this.props;
    this.state = {
      query: '',
      index: null,

      page: 0,
      hitsPerPage: 1,
      totalCount: 0,

      items: [],
      itemsLoaded: false,

      availableTags,
      selectedTags: [],

      showLoadMoreButton: false,
    };
  }

  componentDidMount() {
    const index = setupAlgoliaIndex({
      containerId: 'history',
      indexName: 'UserHistory',
    });

    // get default result set from Algolia
    const { hitsPerPage } = this.state;
    index.search('', { hitsPerPage }).then(content => {
      this.setState({
        items: content.hits,
        totalCount: content.nbHits,
        index,
        itemsLoaded: true,
        showLoadMoreButton: content.hits.length === hitsPerPage,
      });
    });
  }

  handleTyping = event => {
    const query = event.target.value;
    const { selectedTags } = this.state;

    this.setState({ page: 0, items: [] });
    this.search(query, { tags: selectedTags });
  };

  toggleTag = (event, tag) => {
    event.preventDefault();

    const { query, selectedTags } = this.state;
    const newTags = selectedTags;
    if (newTags.indexOf(tag) === -1) {
      newTags.push(tag);
    } else {
      newTags.splice(newTags.indexOf(tag), 1);
    }

    this.setState({ selectedTags: newTags, page: 0, items: [] });
    this.search(query, { tags: newTags });
  };

  loadNextPage = () => {
    const { query, selectedTags, page } = this.state;
    this.setState({ page: page + 1 });
    this.search(query, { selectedTags });
  };

  search(query, { tags }) {
    const { index, hitsPerPage, page, items } = this.state;
    const filters = { hitsPerPage, page };

    if (tags && tags.length > 0) {
      filters.tagFilters = tags;
    }

    index.search(query, filters).then(content => {
      const allItems = [...items, ...content.hits];

      this.setState({
        query,
        items: allItems,
        totalCount: content.nbHits,
        showLoadMoreButton: content.hits.length === hitsPerPage,
      });
    });
  }

  renderNoItems() {
    const { selectedTags, query } = this.state;

    return (
      <div className="history-empty">
        <h1>
          {selectedTags.length === 0 && query.length === 0
            ? 'Your History is Lonely'
            : 'Nothing with this filter ðŸ¤”'}
        </h1>
      </div>
    );
  }

  renderItems() {
    const { items, itemsLoaded } = this.state;

    if (items.length === 0 && itemsLoaded) {
      return this.renderNoItems();
    }

    return items.map(item => (
      <div className="history-item-wrapper">
        <a className="history-item" href={item.article_path}>
          <div className="history-item-title">{item.article_title}</div>

          <div className="history-item-details">
            <a
              className="history-item-user"
              href={`/${item.article_user.username}`}
            >
              <img src={item.article_user.profile_image_90} alt="Profile Pic" />
              {item.article_user.name}
ãƒ»
              {item.article_reading_time}
              {' '}
min readãƒ»
              {`visited on ${item.readable_visited_at}`}
ãƒ»
            </a>
            <span className="history-item-tag-collection">
              {item.article_tags.map(tag => (
                <a className="history-item-tag" href={`/t/${tag}`}>
                  #
                  {tag}
                </a>
              ))}
            </span>
          </div>
        </a>
      </div>
    ));
  }

  render() {
    const {
      itemsLoaded,
      totalCount,
      availableTags,
      selectedTags,
      showLoadMoreButton,
    } = this.state;

    const allItems = this.renderItems();

    return (
      <div className="home history-home">
        <div className="side-bar">
          <div className="widget history-filters">
            <input onKeyUp={this.handleTyping} placeHolder="search your list" />
            <ItemListTags
              availableTags={availableTags}
              selectedTags={selectedTags}
              onClick={this.toggleTag}
            />
          </div>
        </div>
        <div className="items-container">
          <div
            className={`history-results ${
              itemsLoaded ? 'history-results--loaded' : ''
            }`}
          >
            <div className="history-results-header">
              History
              {` (${totalCount})`}
            </div>
            <div>{allItems}</div>
          </div>
          <ItemListLoadMoreButton
            show={showLoadMoreButton}
            onClick={this.loadNextPage}
          />
        </div>
      </div>
    );
  }
}

History.propTypes = {
  availableTags: PropTypes.arrayOf(PropTypes.string).isRequired,
};
