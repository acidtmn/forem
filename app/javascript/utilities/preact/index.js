import { render as preactRender } from 'preact';
import { unmountComponentAtNode } from 'preact/compat';

/**
 * Renders a Preact component into the given parent DOM element.
 *
 * Note: This function is calling Preact's render under the hood along with
 * a call to InstantClick's change event that explicitly unmounts the component rendered.
 *
 * @param {ComponentChild} vnode
 * @param {Element | Document | ShadowRoot | DocumentFragment} parent
 * @param {Element | Text} replaceNode
 */
export function render(vnode, parent, replaceNode) {
  InstantClick.on('change', () => {
    parent && unmountComponentAtNode(parent);
  });

  preactRender(vnode, parent, replaceNode);
}
