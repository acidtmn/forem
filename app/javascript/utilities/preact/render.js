import { render as preactRender } from 'preact';
import { unmountComponentAtNode } from 'preact/compat';

/**
 * Renders a Preact component into the given parent DOM element.
 *
 * Note: This function is calling Preact's render under the hood along with
 * a call to InstantClick's change event that explicitly unmounts the component rendered.
 *
 * @param {object} vnode The virtual DOM node representing a component.
 * @param {Element | Document | ShadowRoot | DocumentFragment} parent The DOM element where the component will render.
 * @param {Element | Text} replaceNode An optional DOM element that must of the given parent.
 * Instead of inferring where to start rendering, it will update or replace the passed element using Preact's diffing algorithm.
 */
export function render(vnode, parent, replaceNode) {
  InstantClick.on('change', () => {
    // We need to explicitly unmount a Preact component since we are using InstantClick. Failing to do so
    // would leave references to past components.
    parent && unmountComponentAtNode(parent);
  });

  preactRender(vnode, parent, replaceNode);
}
