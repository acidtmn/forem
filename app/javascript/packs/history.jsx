import { h, render } from 'preact';
import { History } from '../history/history';

function loadElement() {
  const root = document.getElementById('history');
  if (root) {
    render(<History />, root, root.firstElementChild);
  }
}

window.InstantClick.on('change', () => {
  loadElement();
});

loadElement();
