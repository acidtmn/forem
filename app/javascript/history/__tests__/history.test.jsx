import { h } from 'preact';
import render from 'preact-render-to-json';
import { JSDOM } from 'jsdom';
import fetch from 'jest-fetch-mock';
import { shallow } from 'preact-render-spy';
import { History } from '../history';
import algoliasearch from '../__mocks__/algoliasearch';

global.fetch = fetch;
const doc = new JSDOM('<!doctype html><html><body></body></html>');
global.document = doc;
global.window = doc.defaultView;
global.window.algoliasearch = algoliasearch;
global.window.currentUser = { id: 'modID' };

describe('<History />', () => {
  describe('itemsLoaded', () => {
    describe('when there are Loaded Items', () => {
      it('adds "results--loaded" class', () => {
        const history = shallow(
          <History availableTags={['discuss', 'javascript']} />,
        );

        expect(history.find('.results--loaded').exists()).toEqual(true);
      });
    });

    describe('when there are no Loaded Items', () => {
      it('keeps default class', () => {
        const history = shallow(
          <History availableTags={['discuss', 'javascript']} />,
        );

        expect(history.find('.results--loaded').exists()).toEqual(false);
      });
    });
  });

  describe('totalCount', () => {
    describe(' when totalCount equals zero', () => {
      it('results-header inner html equals empty', () => {
        const history = shallow(
          <History availableTags={['discuss', 'javascript']} />,
        );

        expect(history.find('.results-header').text()).toEqual(
          'History (empty)',
        );
      });
    });

    describe(' when totalCount is bigger than zero', () => {
      it('results-header inner html equals a list', () => {
        const history = shallow(
          <History availableTags={['discuss', 'javascript']} />,
        );

        expect(history.find('.results-header').text()).toEqual('History ( )');
      });
    });
  });

  describe('itemsToRender', () => {
    describe('When there are items to render', () => {
      it('render items', () => {
        expect(true).toBe(true);
      });
    });
  });

  it('renders properly', () => {
    const tree = render(<History availableTags={['discuss']} />);
    expect(tree).toMatchSnapshot();
  });
});
