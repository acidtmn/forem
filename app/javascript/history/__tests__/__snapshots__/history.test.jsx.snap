// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<History /> renders properly 1`] = `
<div
  class="home history-home"
>
  <div
    class="side-bar"
  >
    <div
      class="widget history-filters"
    >
      <input
        onKeyUp={[Function]}
        placeHolder="search your list"
      />
      <div
        class="history-tags"
      >
        <a
          class="history-tag "
          data-no-instant={true}
          href="/t/discuss"
          onClick={[Function]}
        >
          #discuss
        </a>
      </div>
    </div>
  </div>
  <div
    class="history-results "
  >
    <div
      class="history-results-header"
    >
      History (0)
    </div>
    <div />
  </div>
</div>
`;
